cli/admin/src/main/java/io/pravega/cli/admin/dataRecovery/SegmentReader.java

```diff
@@ -0,0 +1,177 @@
+            throw new Exception("Failed to create file " + f.getAbsolutePath());
+        }
+        this.fileOutputStream = new FileOutputStream(f.getAbsolutePath(), true);
+        outputInfo("Created file '%s'", f.getAbsolutePath());
+    }
+
+    @Override
+    public void execute() throws Exception {
+        @Cleanup
+        Storage storage = this.storageFactory.createStorageAdapter();
+
+        // Get the storage using the config.
+        storage.initialize(CONTAINER_EPOCH);
+        outputInfo("Loaded %s Storage.", getServiceConfig().getStorageImplementation().toString());
+
+        // create CSV files for listing the segments and their details
+        createFileAndDirectory();
+
+        outputInfo("Writing segment's content to the file...");
+        int countEvents = readSegment(storage, this.segmentName, TIMEOUT);
+        outputInfo("Number of events found = %d.", countEvents);
+
+        outputInfo("Closing the file...");
+        this.fileOutputStream.close();
+
+        outputInfo("The segment's contents have been written to the file.");
+        outputInfo("Path to the file '%s'", this.filePath);
+        outputInfo("Done!");
+    }
+
+    /**
+     * Reads the contents of a given segment.
+     * @param storage                   A storage instance to read the segment's contents.
+     * @param segmentName               The name of the segment.
+     * @param timeout                   A timeout for the operation.
+     * @throws Exception                If an exception occurred.
+     * @return                          Number of events found.
+     */
+    private int readSegment(Storage storage, String segmentName, Duration timeout)
+            throws Exception {
+
+        val segmentInfo = storage.getStreamSegmentInfo(segmentName, timeout).get(timeout.toMillis(), TimeUnit.MILLISECONDS);
+        int bytesToRead = (int) segmentInfo.getLength();
+        val sourceHandle = storage.openRead(segmentName).get(timeout.toMillis(), TimeUnit.MILLISECONDS);
+
+        byte[] header; // stores header
+        byte[] data; // stores data
+
+        int countEvents = 0; // Keeps count of events
+        int offset = 0;     // Moves the offset to read a new chunk less then equal to BUFFER_SIZE
+
+        byte[] buffer = new byte[BUFFER_SIZE];
+        int runnerOffset;
+        while (offset < bytesToRead) { // read till offset reaches the end
+            int size = storage.read(sourceHandle, offset, buffer, 0, Math.min(BUFFER_SIZE, bytesToRead - offset), timeout)
+                    .get(timeout.toMillis(), TimeUnit.MILLISECONDS);
+            outputInfo("Read size: %d", size);
+
+            runnerOffset = 0; // to read the current buffer
+            while ( (runnerOffset + 8) < size) { // It should be less, unless there can be 0 bytes event
+                header = Arrays.copyOfRange(buffer, runnerOffset, runnerOffset + 8);
+
+                long length = convertByteArrayToLong(header);
```

Comment:
Also I don't understand why do we need to to do this event-by-event? Why can't we just dump the whole segment into the file? It will be much faster and plus, not all segments contain events (think table segments).